riscv32: LiteX VexRiscv support

mateusz-holenko opened the pull request:
This commit adds support for LiteX with VexRiscv softcore CPU, running
on Digilent ARTY development board, alongside with UART driver and
interrupt controller for LiteX and VexRiscv.

galak reviewed (changes requested):
Can you break this up into separate commits:

Each driver as one commit, the soc code as one commit, and the board code as one.

Thanks.

galak commented on boards/riscv32/arty_litex_vexriscv/Kconfig.board:
```diff
@@ -0,0 +1,4 @@
+config BOARD_ARTY_LITEX_VEXRISCV
+	bool "Digilent Arty dev board with LiteX/VexRiscV CPU"
+	select HAS_DTS
```
The select HAS_DTS shouldn't be needed.  Should already be set for any RISC-V 32 port.

mateusz-holenko replied:
You are right, we checked - it's not needed and will be removed.

mateusz-holenko replied:
It's done.

galak commented on boards/riscv32/arty_litex_vexriscv/arty_litex_vexriscv.yaml:
```diff
@@ -0,0 +1,10 @@
+arch: riscv32
+toolchain:
+  - zephyr
+testing:
```
Would be good to add the amount of memory / flash here.

mateusz-holenko replied:
There is no flash on the board, but we will add the ram size.

galak commented on drivers/interrupt_controller/vexriscv_litex.c:
```diff
@@ -0,0 +1,100 @@
+/*
+ * Copyright (c) 2019 Antmicro <www.antmicro.com>
```
Is this copyright correct?

mateusz-holenko replied:
Do you mean year? The WIP version of this code was indeed published in 2018, so I'll update this.

mateusz-holenko replied:
I changed copyright year to 2018-2019 in every file.

galak commented on soc/riscv32/litex-vexriscv/dts_fixup.h:
```diff
@@ -0,0 +1,22 @@
+ */
+
+/* UART 0 */
+#define CONFIG_UART_0_BASE_ADDR		DT_LITEX_UART0_E0001800_BASE_ADDRESS
```
We moved away from using CONFIG_ prefixed defines.  I'd suggest maybe seeing if you can just use the standard generated defines instead of needing a dts_fixup.h.

mateusz-holenko replied:
We'll use the generated defines directly.

mateusz-holenko replied:
It's done.

galak reviewed (changes requested):
Some various clean up / questions.

dbkinder commented on boards/riscv32/arty_litex_vexriscv/doc/arty_litex_vexriscv.rst:
```diff
@@ -0,0 +1,33 @@
+Booting
+=======
+
+You can boot from serial port using `flterm: <https://github.com/timvideos/flterm>`_, e.g.:
```
First I've heard of this tool (flterm).  Does it do something that other serial port tools we typically mention don't do?

pgielda replied:
LiteX (SoC) has a built-in ROM bootloader ("bios") that supports uploading binaries but a special tool (e.g. flterm) is needed.

dbkinder reviewed (approved):
Docs look OK.  I'll let the developers comment on use of the flterm tools...

fkokosinski committed:
drivers: serial: add LiteUART driver

Add LiteX UART driver with bindings for this device.

Signed-off-by: Filip Kokosinski <fkokosinski@internships.antmicro.com>
Signed-off-by: Mateusz Holenko <mholenko@antmicro.com>

fkokosinski committed:
drivers: timer: add LiteX timer driver

Add LiteX timer driver with bindings for this device.

Signed-off-by: Filip Kokosinski <fkokosinski@internships.antmicro.com>
Signed-off-by: Mateusz Holenko <mholenko@antmicro.com>

fkokosinski committed:
drivers: interrupt_controller: add LiteX interrupt controller driver

Add LiteX interrupt controller driver and bindings for this device.

Signed-off-by: Filip Kokosinski <fkokosinski@internships.antmicro.com>
Signed-off-by: Mateusz Holenko <mholenko@antmicro.com>

fkokosinski committed:
soc: riscv32: add LiteX VexRiscV SoC

Add LiteX with softcore CPU VexRiscV SoC definitions and default
configurations.

Signed-off-by: Filip Kokosinski <fkokosinski@internships.antmicro.com>
Signed-off-by: Mateusz Holenko <mholenko@antmicro.com>

fkokosinski committed:
boards: riscv32: add LiteX VexRiscV board

Add LiteX VexRiscV board platform definitions and
default configurations.

Signed-off-by: Filip Kokosinski <fkokosinski@internships.antmicro.com>
Signed-off-by: Mateusz Holenko <mholenko@antmicro.com>

enjoy-digital commented:
To ease support for others boards/SoCs in the future, i don't think the board directory should be 
`arty_litex_vexriscv`but maybe `litex-vexriscv` as you are doing for `soc/riscv32/litex-vexriscv/`.

Different LiteX generated SoCs can have the same exact parameters from Zephyr point of view, so it would be interesting to minimize information related to the boards. (only a few parameters will change between boards/SoCs).

kgugala commented:
@enjoy-digital I agree, this could be a generic LiteX-VexRiscV port. The only thing arty dependant is RAM. This number is determined by the DDR chips on the board. Maybe we can reorganize the code so it adds generic LiteX support and only dts for Arty.

enjoy-digital commented:
@kgugala: yes that seems fine.

mateusz-holenko commented:
@galak I addressed all your questions, do you have any other comments?

mateusz-holenko commented:
When working on @enjoy-digital's suggestion for reorganizing the code to add generic LiteX support, we realized it would be beneficial to move memory definitions from KConfig to DTS for riscv32 boards.

We will create a separate PR with that change and wait for it to be merged before carrying on with this one.

mateusz-holenko commented:
A link to the above-mentioned PR: #15418.

mateusz-holenko commented:
Since the #15418 is merged to master, we can proceed with this PR.

@enjoy-digital As you suggested, the platform has been renamed to `litex_vexriscv`. There is a separate dts file that includes SoC and adds memory definition, but we decided to drop the reference to Arty board making the platform a generic LiteX.

@galak I have rebased it onto the current master, addressed all your comments. Are we ready to merge it?

enjoy-digital commented:
@mateusz-holenko: great, thanks.

galak commented on soc/riscv32/litex-vexriscv/Kconfig.defconfig:
```diff
@@ -0,0 +1,31 @@
+	default 12
+
+# currently n-selecting, because of UART Tx interrupt related bug
+config SERIAL_SUPPORT_INTERRUPT
```
I assume from the comment you are saying that interrupt's aren't supported.  Is this really accomplishing what you want?

mateusz-holenko replied:
You are right, that was done wrong. I removed this line from soc-level Kconfig and added `depends on !SERIAL_SUPPORT_INTERRUPT` in uart-level Kconfig.

galak commented on soc/riscv32/litex-vexriscv/soc.h:
```diff
@@ -0,0 +1,20 @@
+#include <generated_dts_board.h>
+
+/* MIE address */
+#define SOC_MIE_IEN                  0x880
```
Is this define need anywhere but the IRQ code?  If not, just move it there.

mateusz-holenko replied:
I got rid of this define at all.

galak reviewed (changes requested):
Some minor cleanups.

mateusz-holenko commented on dts/riscv32/riscv32-litex-vexriscv.dtsi:
```diff
@@ -0,0 +1,58 @@
+			compatible = "vexriscv,intc0";
+			interrupt-controller;
+			reg = <0xbc0 0x4 0xfc0 0x4>;
+			reg-names = "control";
```
I set proper names in `reg-names` and referenced to them directly in the driver.

mateusz-holenko commented on dts/riscv32/riscv32-litex-vexriscv.dtsi:
```diff
@@ -0,0 +1,58 @@
+		#size-cells = <0>;
+		cpu@0 {
+			clock-frequency = <100000000>;
+			compatible = "litex", "riscv";
```
I changed the name to be more specific. VexRiscv is an independent core that can be used outside LiteX, so I set compatible to `spinalhdl,vexriscv` and left the second value `riscv` as I saw this is the pattern for other RISC-V cores.

mateusz-holenko commented:
@galak I addressed all comments.

galak commented on CODEOWNERS:
```diff
@@ -170,6 +175,9 @@
 /dts/bindings/*/st*                       @erwango
 /dts/bindings/sensor/ams*                 @alexanderwachter
 /dts/bindings/*/sifive*                   @mateusz-holenko @kgugala @pgielda @nategraff-sifive
+/dts/bindings/*/vexriscv,intc0.yaml       @mateusz-holenko @kgugala @pgielda
```
Can you change this to be:
```
/dts/bindings/*/vexrisc*
/dts/bindings/*/litex*
```

there appears to be some bug if we have `,` in this file.

mateusz-holenko replied:
@galak I refactored CODEOWNERS as suggested.

galak reviewed (changes requested):
need a minor change in CODEOWNERS due to a github bug.

mgielda commented:
OK, @galak can you merge?
